File: src/orchestrator/risk-policy.ts
import { ExecutionPlan } from './protocols/types';

export type RiskLevel = 'low' | 'medium' | 'high';
export type RiskPath = 'light' | 'standard' | 'full';
export type VerificationLevel = 'none' | 'basic' | 'full';

export interface RiskAssessment {
  level: RiskLevel;
  path: RiskPath;
  hardStop: boolean;
  verification: VerificationLevel;
  score: number;
  signals: string[];
}

/**
 * RiskPolicy
 * 统一风险评估与路径选择的最小策略内核。
 */
export class RiskPolicy {
  evaluate(prompt: string, plan: ExecutionPlan): RiskAssessment {
    const signals: string[] = [];
    let score = 0;
    const subTasks = plan.subTasks || [];
    const acceptanceCount = plan.acceptanceCriteria?.length || 0;
    const featureLen = (plan.featureContract || '').length;
    const promptLen = (prompt || '').length;

    const targetFiles = new Set<string>();
    for (const task of subTasks) {
      for (const file of task.targetFiles || []) {
        if (file) targetFiles.add(file);
      }
    }

    const fileCount = targetFiles.size > 0 ? targetFiles.size : subTasks.length;
    const moduleCount = this.countModules(targetFiles);

    const fileScore = fileCount <= 0 ? 0 : fileCount <= 2 ? 1 : fileCount <= 5 ? 2 : 3;
    if (fileScore > 0) {
      score += fileScore * 2;
      signals.push(`file_count_${fileScore}`);
    }

    const moduleScore = moduleCount === 0 ? 0 : moduleCount === 1 ? 1 : 2;
    if (moduleScore > 0) {
      score += moduleScore * 3;
      signals.push(`module_count_${moduleScore}`);
    }

    if (this.hasInterfaceChange(prompt, plan)) {
      score += 3 * 4;
      signals.push('interface_change');
    }

    if (this.hasConfigChange(targetFiles)) {
      score += 3 * 4;
      signals.push('config_or_dependency_change');
    }

    const failureRate = this.getFailureRate(plan);
    if (failureRate >= 0) {
      const failureScore = failureRate > 0.3 ? 2 : failureRate >= 0.1 ? 1 : 0;
      if (failureScore > 0) {
        score += failureScore * 2;
        signals.push(`failure_rate_${failureScore}`);
      }
    }

    if (targetFiles.size === 0 && subTasks.length > 1) {
      score += 6;
      signals.push('unknown_file_scope');
    }

    let level: RiskLevel = 'low';
    if (score >= 13) level = 'high';
    else if (score >= 7) level = 'medium';

    const path: RiskPath = level === 'low' ? 'light' : level === 'medium' ? 'standard' : 'full';
    const hardStop = level !== 'low';
    const verification: VerificationLevel = level === 'high' ? 'full' : 'basic';

    return { level, path, hardStop, verification, score, signals };
  }

  private countModules(files: Set<string>): number {
    if (files.size === 0) return 0;
    const modules = new Set<string>();
    for (const file of files) {
      const normalized = file.replace(/\\/g, '/').replace(/^\.\//, '');
      const segment = normalized.split('/')[0];
      if (segment) modules.add(segment);
    }
    return modules.size;
  }

  private hasInterfaceChange(prompt: string, plan: ExecutionPlan): boolean {
    const combined = [prompt, plan.analysis, plan.featureContract].filter(Boolean).join('\n');
    if (!combined) return false;
    const keywords = [
      'API',
      '接口',
      'endpoint',
      'schema',
      '契约',
      '请求',
      '响应',
      '字段',
      'payload',
    ];
    return keywords.some(keyword => combined.includes(keyword));
  }

  private hasConfigChange(files: Set<string>): boolean {
    if (files.size === 0) return false;
    const configFiles = [
      'package.json',
      'package-lock.json',
      'pnpm-lock.yaml',
      'yarn.lock',
      'requirements.txt',
      'pyproject.toml',
      'Pipfile',
      'go.mod',
      'go.sum',
      'Cargo.toml',
      'Cargo.lock',
      'pom.xml',
      'build.gradle',
      'build.gradle.kts',
      'tsconfig.json',
      'vite.config',
      'webpack.config',
      'next.config',
    ];
    for (const file of files) {
      const base = file.split('/').pop() || '';
      if (configFiles.some(cfg => base === cfg || base.startsWith(cfg))) {
        return true;
      }
    }
    return false;
  }

  private getFailureRate(plan: ExecutionPlan): number {
    const maybeRate = (plan as { failureRate?: number }).failureRate;
    if (typeof maybeRate === 'number' && !Number.isNaN(maybeRate)) {
      return maybeRate;
    }
    return -1;
  }
}

File: crates/magi-orchestrator/src/risk_policy.rs
use magi_core::RiskLevel;
use serde::{Deserialize, Serialize};
use std::collections::HashSet;

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum RiskPath {
    Light,
    Standard,
    Full,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum VerificationLevel {
    None,
    Basic,
    Full,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RiskAssessment {
    pub level: RiskLevel,
    pub path: RiskPath,
    pub hard_stop: bool,
    pub verification: VerificationLevel,
    pub score: u32,
    pub signals: Vec<String>,
}

pub struct RiskPolicyInput<'a> {
    pub prompt: &'a str,
    pub analysis: Option<&'a str>,
    pub feature_contract: Option<&'a str>,
    pub sub_task_count: usize,
    pub target_files: &'a [String],
    pub acceptance_criteria_count: usize,
    pub failure_rate: Option<f64>,
}

pub fn evaluate_risk(input: &RiskPolicyInput) -> RiskAssessment {
    let mut signals = Vec::new();
    let mut score: u32 = 0;

    let target_files: HashSet<&str> = input.target_files.iter().map(|s| s.as_str()).collect();
    let file_count = if target_files.is_empty() {
        input.sub_task_count
    } else {
        target_files.len()
    };

    let file_score = if file_count == 0 {
        0
    } else if file_count <= 2 {
        1
    } else if file_count <= 5 {
        2
    } else {
        3
    };
    if file_score > 0 {
        score += file_score * 2;
        signals.push(format!("file_count_{file_score}"));
    }

    let module_count = count_modules(&target_files);
    let module_score = if module_count == 0 {
        0
    } else if module_count == 1 {
        1
    } else {
        2
    };
    if module_score > 0 {
        score += module_score * 3;
        signals.push(format!("module_count_{module_score}"));
    }

    if has_interface_change(input.prompt, input.analysis, input.feature_contract) {
        score += 3 * 4;
        signals.push("interface_change".to_string());
    }

    if has_config_change(&target_files) {
        score += 3 * 4;
        signals.push("config_or_dependency_change".to_string());
    }

    if let Some(failure_rate) = input.failure_rate
        && failure_rate >= 0.0
    {
        let failure_score = if failure_rate > 0.3 {
            2
        } else if failure_rate >= 0.1 {
            1
        } else {
            0
        };
        if failure_score > 0 {
            score += failure_score * 2;
            signals.push(format!("failure_rate_{failure_score}"));
        }
    }

    if target_files.is_empty() && input.sub_task_count > 1 {
        score += 6;
        signals.push("unknown_file_scope".to_string());
    }

    let level = if score >= 13 {
        RiskLevel::High
    } else if score >= 7 {
        RiskLevel::Medium
    } else {
        RiskLevel::Low
    };

    let path = match level {
        RiskLevel::Low => RiskPath::Light,
        RiskLevel::Medium => RiskPath::Standard,
        RiskLevel::High => RiskPath::Full,
    };

    let hard_stop = level != RiskLevel::Low;
    let verification = if level == RiskLevel::High {
        VerificationLevel::Full
    } else {
        VerificationLevel::Basic
    };

    RiskAssessment {
        level,
        path,
        hard_stop,
        verification,
        score,
        signals,
    }
}

fn count_modules(files: &HashSet<&str>) -> usize {
    if files.is_empty() {
        return 0;
    }
    let mut modules = HashSet::new();
    for file in files {
        let normalized = file.replace('\\', "/");
        let normalized = normalized.strip_prefix("./").unwrap_or(&normalized);
        if let Some(segment) = normalized.split('/').next()
            && !segment.is_empty()
        {
            modules.insert(segment.to_string());
        }
    }
    modules.len()
}

fn has_interface_change(
    prompt: &str,
    analysis: Option<&str>,
    feature_contract: Option<&str>,
) -> bool {
    let combined = format!(
        "{}\n{}\n{}",
        prompt,
        analysis.unwrap_or(""),
        feature_contract.unwrap_or("")
    );
    if combined.trim().is_empty() {
        return false;
    }
    const KEYWORDS: &[&str] = &[
        "API", "接口", "endpoint", "schema", "契约", "请求", "响应", "字段", "payload",
    ];
    KEYWORDS.iter().any(|kw| combined.contains(kw))
}

fn has_config_change(files: &HashSet<&str>) -> bool {
    if files.is_empty() {
        return false;
    }
    const CONFIG_FILES: &[&str] = &[
        "package.json",
        "package-lock.json",
        "pnpm-lock.yaml",
        "yarn.lock",
        "requirements.txt",
        "pyproject.toml",
        "Pipfile",
        "go.mod",
        "go.sum",
        "Cargo.toml",
        "Cargo.lock",
        "pom.xml",
        "build.gradle",
        "build.gradle.kts",
        "tsconfig.json",
        "vite.config",
        "webpack.config",
        "next.config",
    ];
    for file in files {
        let base = file.rsplit('/').next().unwrap_or(file);
        if CONFIG_FILES
            .iter()
            .any(|cfg| base == *cfg || base.starts_with(cfg))
        {
            return true;
        }
    }
    false
}
